tokio-util/src/sync/poll_semaphore.rs

```diff
@@ -12,17 +12,15 @@ use super::ReusableBoxFuture;
 /// [`Semaphore`]: tokio::sync::Semaphore
 pub struct PollSemaphore {
     semaphore: Arc<Semaphore>,
-    permit_fut: ReusableBoxFuture<Result<OwnedSemaphorePermit, AcquireError>>,
+    permit_fut: Option<ReusableBoxFuture<Result<OwnedSemaphorePermit, AcquireError>>>,
```

Comment:
Another alternative that sometimes works is to initialize the `ReusableBoxFuture` with a zero-sized future, which also would not allocate. But it probably doesn't help much in this specific case.

Reply:
Yup! But that wouldn't let us fast-path `.try_acquire` anymore without an explicit flag, would it?

Reply:
Yeah, I don't think it would help here, but it could be done for [`PollSender`][1], and maybe also the watch/broadcast wrappers in tokio-stream.

[1]: https://docs.rs/tokio-util/0.6/tokio_util/sync/struct.PollSender.html